File: autocomplete_trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        word = word.lower()
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_word = True

    def search_prefix(self, prefix):
        prefix = prefix.lower()
        node = self.root
        for char in prefix:
            if char not in node.children:
                return None
            node = node.children[char]
        return node

    def autocomplete(self, prefix, max_results=10):
        prefix = prefix.lower()
        results = []
        node = self.search_prefix(prefix)
        if not node:
            return []
        self.dfs(node, prefix, results, max_results)
        return results

    def dfs(self, node, prefix, results, max_results):
        if len(results) >= max_results:
            return
        if node.is_end_of_word:
            results.append(prefix)
        for char, child_node in node.children.items():
            self.dfs(child_node, prefix + char, results, max_results)


def load_words_from_file(filename):
    try:
        with open(filename, 'r') as file:
            return [line.strip().lower() for line in file if line.strip()]
    except FileNotFoundError:
        print(f"File '{filename}' not found. Using default sample words.")
        return [
            "apple", "application", "apply", "appliance",
            "banana", "band", "banner",
            "cat", "caterpillar", "cattle",
            "dog", "dodge", "dolphin"
        ]


if __name__ == "__main__":
    words = load_words_from_file('google-10000-english.txt')
    trie = Trie()

    for word in words:
        trie.insert(word)

    print("Trie has been populated with words.")

    while True:
        command = input("\nEnter a prefix, 'reload' to reload words, or 'exit' to quit: ").strip()
        if command.lower() == "exit":
            print("Exiting the program. Goodbye!")
            break
        elif command.lower() == "reload":
            words = load_words_from_file('google-10000-english.txt')
            trie = Trie()
            for word in words:
                trie.insert(word)
            print("Words have been reloaded into the Trie.")
        elif command:
            suggestions = trie.autocomplete(command, max_results=10)
            if suggestions:
                print("Suggestions:", suggestions)
            else:
                print("No suggestions found.")
        else:
            print("Please enter a valid command.")

File: autocomplete_trie.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define ALPHABET_SIZE 26
#define MAX_RESULTS 10
#define MAX_WORD_LENGTH 100

// Trie Node Definition
typedef struct TrieNode {
    struct TrieNode* children[ALPHABET_SIZE];
    int is_end_of_word;
} TrieNode;

// Create a new Trie Node
TrieNode* createNode() {
    TrieNode* node = (TrieNode*)malloc(sizeof(TrieNode));
    if (!node) {
        perror("Failed to allocate memory for TrieNode");
        exit(EXIT_FAILURE);
    }
    node->is_end_of_word = 0;
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        node->children[i] = NULL;
    }
    return node;
}

// Insert a word into the Trie
void insert(TrieNode* root, const char* word) {
    TrieNode* current = root;
    while (*word) {
        char c = tolower(*word) - 'a';
        if (c < 0 || c >= ALPHABET_SIZE) {
            word++;
            continue; // Skip non-alphabet characters
        }
        if (!current->children[c]) {
            current->children[c] = createNode();
        }
        current = current->children[c];
        word++;
    }
    current->is_end_of_word = 1;
}

// Search for a prefix in the Trie
TrieNode* searchPrefix(TrieNode* root, const char* prefix) {
    TrieNode* current = root;
    while (*prefix) {
        char c = tolower(*prefix) - 'a';
        if (c < 0 || c >= ALPHABET_SIZE || !current->children[c]) {
            return NULL;
        }
        current = current->children[c];
        prefix++;
    }
    return current;
}

// Depth-First Search to collect words
void dfs(TrieNode* node, char* prefix, int prefix_length, char results[][MAX_WORD_LENGTH], int* result_count, int max_results) {
    if (*result_count >= max_results || !node) {
        return;
    }
    if (node->is_end_of_word) {
        prefix[prefix_length] = '\0';
        strcpy(results[*result_count], prefix);
        (*result_count)++;
    }
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        if (node->children[i]) {
            prefix[prefix_length] = 'a' + i;
            dfs(node->children[i], prefix, prefix_length + 1, results, result_count, max_results);
        }
    }
}

// Autocomplete function
void autocomplete(TrieNode* root, const char* prefix, char results[][MAX_WORD_LENGTH], int* result_count) {
    TrieNode* node = searchPrefix(root, prefix);
    if (!node) {
        *result_count = 0;
        return;
    }
    char buffer[MAX_WORD_LENGTH];
    strcpy(buffer, prefix);
    dfs(node, buffer, strlen(prefix), results, result_count, MAX_RESULTS);
}

// Load words from a file into the Trie
int loadWordsFromFile(const char* filename, TrieNode* root) {
    FILE* file = fopen(filename, "r");
    if (!file) {
        perror("File not found. Using default words.");
        return 0;
    }
    char word[MAX_WORD_LENGTH];
    while (fgets(word, sizeof(word), file)) {
        word[strcspn(word, "\n")] = '\0'; // Remove newline character
        insert(root, word);
    }
    fclose(file);
    return 1;
}

// Free the Trie
void freeTrie(TrieNode* node) {
    if (!node) return;
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        freeTrie(node->children[i]);
    }
    free(node);
}
